server/stream.go

```diff
@@ -780,11 +782,16 @@ func (mset *Stream) storeUpdates(md, bd int64, seq uint64, subj string) {
 	// If we have a single negative update then we will process our consumers for stream pending.
 	// Purge and Store handled separately inside individual calls.
 	if md == -1 {
-		mset.mu.RLock()
+		needLock := atomic.LoadInt32(&mset.lh) == 0
```

Comment:
I think this is wrong. If 2 threads are doing store operations that lead to the invocation of this callback, one thread may be under the mset lock and has stored the atomic, while the other thread is not under the mset lock. If the second thread evaluates the atomic it will think that it does not need the lock - because it has been set by the first thread. It is possible that the first thread finishes processing of callback, unsets atomic and releases lock, and then the second thread resumes execution and proceeds with executing this code without the lock, which is technically wrong. (granted, it is unlikely in this sequence of events since it means that thread 2 execution would have to be suspended after evaluation of the atomic and before getting the mset RLock, but...)

So either all invocations of this callback would have to be made under the mset lock,  which clearly we don't always (like Purge, etc..), or if there is no need for high accuracy, then these callback executions could be serialized and dispatched from separate long running go routine (similar to async callbacks in go client).

NOTE: maybe we could ensure that cases where `md==-1` is invoked from code that does have the mset lock, in which case we would not need this code at all. I am guessing that the case is for removing of a message or purge when there was a single message, no?

Reply:
The atomic is only set when the mset lock is held and is also only cleared before said lock is released. Yes this happens when we have a limit on number of messages or bytes and storing the new message causes us to drop one.

Reply:
If we do not protect the store with the mset lock though I could see a callback here from a RemoveMsg() that may cause issues. We could make sure to protect and single removals.

What I wanted to avoid is releasing lock and reacquiring since I need those preconditions set before I release the lock.

Reply:
I understand that the atomic is set under the mset lock, the issue is that a thread could be doing say a Purge() which is executed outside of the mset lock, which another thread may be processing inbound msg and calling StoreMsg() (where this atomic is set/unset) under the mset lock.

Reply:
Purge() is ok. Its RemoveMsg() we need to make sure is protected by mset lock.

Reply:
https://github.com/nats-io/nats-server/pull/1760#discussion_r534440035

Right, this is why I mentioned that we should either always have the storeUpdates callback be invoked from under the mset lock, which means all store operations that can trigger the store update callback should be under the mset lock, or at least look at cases where callback is invoked with md==-1, which I assume is remove msg and purge of single message?

Reply:
Purge() calls with "-(purged)", so if purged is 1, then you have the problem.

Reply:
Let me see if there is a better way. Thanks for the feedback, helpful (and correct)..

Reply:
ok please take another look. Thanks..